File: backend/app/api/v1/planning.py
from __future__ import annotations

from datetime import date
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.core.audit import audit
from app.core.db import get_db
from app.core.deps import get_current_user
from app.models.planning import TaskSeries
from app.models.user import User
from app.schemas.common import OkResponse
from app.schemas.planning import (
    AutoPlanIn,
    AutoPlanOut,
    CalendarDayOut,
    CalendarOut,
    OccurrenceIn,
    RescheduleIn,
    SeriesCreate,
    SeriesOut,
    SeriesUpdate,
    TaskCreate,
    TaskOut,
    TaskUpdate,
    WeekPrintOut,
)
from app.services import activities as activity_service
from app.services import planning as svc

router = APIRouter(tags=["planning"])


def _task_out(t) -> TaskOut:
    return TaskOut(**svc.task_view(t))


# --- Tarefas ---------------------------------------------------------------


@router.get("/tasks", response_model=list[TaskOut])
def list_tasks(
    start: date,
    end: date,
    activity_id: UUID | None = None,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> list[TaskOut]:
    """Tarefas do intervalo (datas locais) com ocorrências virtuais de séries (`id` nulo)."""
    if activity_id is not None:
        activity_service.get_activity(db, user, activity_id)
    return [TaskOut(**v) for v in svc.list_tasks(db, user, start, end, activity_id=activity_id)]


@router.post("/tasks", response_model=TaskOut, status_code=201)
def create_task(
    payload: TaskCreate, db: Session = Depends(get_db), user: User = Depends(get_current_user)
) -> TaskOut:
    act = activity_service.get_activity(db, user, payload.activity_id)
    t = svc.create_task(
        db,
        user,
        act,
        title=payload.title,
        kind=payload.kind,
        local_date=payload.local_date,
        start_time=payload.start_time,
        estimated_seconds=payload.estimated_seconds,
        subject_id=payload.subject_id,
        topic_id=payload.topic_id,
        material_id=payload.material_id,
        page_from=payload.page_from,
        page_to=payload.page_to,
        priority=payload.priority,
        due_date=payload.due_date,
        notes=payload.notes,
        pinned=payload.pinned,
    )
    db.commit()
    return _task_out(t)


# Séries antes de /tasks/{task_id} para não capturar "series" como id.


@router.post("/tasks/series", response_model=SeriesOut, status_code=201)
def create_series(
    payload: SeriesCreate, db: Session = Depends(get_db), user: User = Depends(get_current_user)
) -> SeriesOut:
    act = activity_service.get_activity(db, user, payload.activity_id)
    s = svc.create_series(
        db,
        user,
        act,
        title=payload.title,
        weekdays=payload.weekdays,
        start_date=payload.start_date,
        end_date=payload.end_date,
        start_time=payload.start_time,
        estimated_seconds=payload.estimated_seconds,
        subject_id=payload.subject_id,
        topic_id=payload.topic_id,
        kind=payload.kind,
    )
    audit(db, actor_id=user.id, action="series.create", target_type="series", target_id=str(s.id))
    db.commit()
    return SeriesOut(**svc.series_view(s))


@router.get("/tasks/series", response_model=list[SeriesOut])
def list_series(
    activity_id: UUID | None = None,
    include_inactive: bool = False,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> list[SeriesOut]:
    q = select(TaskSeries).where(TaskSeries.user_id == user.id)
    if activity_id is not None:
        activity_service.get_activity(db, user, activity_id)
        q = q.where(TaskSeries.activity_id == activity_id)
    if not include_inactive:
        q = q.where(TaskSeries.active.is_(True))
    rows = db.execute(q.order_by(TaskSeries.created_at)).scalars()
    return [SeriesOut(**svc.series_view(s)) for s in rows]


@router.patch("/tasks/series/{series_id}", response_model=SeriesOut)
def update_series(
    series_id: UUID,
    payload: SeriesUpdate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> SeriesOut:
    s = svc.get_series(db, user, series_id)
    svc.update_series(db, user, s, payload.model_dump(exclude_unset=True))
    db.commit()
    return SeriesOut(**svc.series_view(s))


@router.delete("/tasks/series/{series_id}", response_model=OkResponse)
def delete_series(
    series_id: UUID, db: Session = Depends(get_db), user: User = Depends(get_current_user)
) -> OkResponse:
    s = svc.get_series(db, user, series_id)
    svc.deactivate_series(db, s)
    audit(
        db, actor_id=user.id, action="series.deactivate", target_type="series", target_id=str(s.id)
    )
    db.commit()
    return OkResponse(message="Série desativada. As ocorrências já editadas continuam no plano.")


@router.post(
    "/tasks/series/{series_id}/occurrences/{local_date}", response_model=TaskOut, status_code=201
)
def materialize_occurrence(
    series_id: UUID,
    local_date: date,
    payload: OccurrenceIn | None = None,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> TaskOut:
    """Persiste a ocorrência (series_id, data) e aplica as alterações enviadas. Idempotente."""
    s = svc.get_series(db, user, series_id)
    p = payload or OccurrenceIn()
    t = svc.materialize_occurrence(db, user, s, local_date, p.model_dump(exclude_unset=True))
    db.commit()
    return _task_out(t)


@router.get("/tasks/{task_id}", response_model=TaskOut)
def get_task(
    task_id: UUID, db: Session = Depends(get_db), user: User = Depends(get_current_user)
) -> TaskOut:
    return _task_out(svc.get_task(db, user, task_id))


@router.patch("/tasks/{task_id}", response_model=TaskOut)
def update_task(
    task_id: UUID,
    payload: TaskUpdate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> TaskOut:
    t = svc.get_task(db, user, task_id)
    fields = payload.model_dump(exclude_unset=True)
    expected = fields.pop("expected_version", None)
    svc.update_task(db, user, t, fields, expected_version=expected)
    db.commit()
    return _task_out(t)


@router.delete("/tasks/{task_id}", response_model=OkResponse)
def delete_task(
    task_id: UUID, db: Session = Depends(get_db), user: User = Depends(get_current_user)
) -> OkResponse:
    t = svc.get_task(db, user, task_id)
    svc.delete_task(db, t)
    db.commit()
    return OkResponse(message="Tarefa excluída.")


@router.post("/tasks/{task_id}/complete", response_model=TaskOut)
def complete_task(
    task_id: UUID, db: Session = Depends(get_db), user: User = Depends(get_current_user)
) -> TaskOut:
    """Marca como concluída. Conclusão não lança minutos: tempo só entra por sessão."""
    t = svc.get_task(db, user, task_id)
    svc.set_status(t, "done")
    db.commit()
    return _task_out(t)


@router.post("/tasks/{task_id}/uncomplete", response_model=TaskOut)
def uncomplete_task(
    task_id: UUID, db: Session = Depends(get_db), user: User = Depends(get_current_user)
) -> TaskOut:
    t = svc.get_task(db, user, task_id)
    svc.set_status(t, "planned")
    db.commit()
    return _task_out(t)


@router.post("/tasks/{task_id}/skip", response_model=TaskOut)
def skip_task(
    task_id: UUID, db: Session = Depends(get_db), user: User = Depends(get_current_user)
) -> TaskOut:
    t = svc.get_task(db, user, task_id)
    svc.set_status(t, "skipped")
    db.commit()
    return _task_out(t)


@router.post("/tasks/{task_id}/reschedule", response_model=TaskOut)
def reschedule_task(
    task_id: UUID,
    payload: RescheduleIn,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> TaskOut:
    """Mantém a mesma tarefa (mesmo id); guarda `original_date` na primeira reprogramação."""
    t = svc.get_task(db, user, task_id)
    svc.reschedule_task(
        db,
        t,
        local_date=payload.local_date,
        start_time=payload.start_time,
        clear_start_time=payload.clear_start_time,
        expected_version=payload.expected_version,
    )
    db.commit()
    return _task_out(t)


# --- Calendário e impressão ------------------------------------------------


@router.get("/calendar", response_model=CalendarOut)
def calendar(
    start: date,
    end: date,
    activity_id: UUID | None = None,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> CalendarOut:
    days = svc.calendar(db, user, start, end, activity_id=activity_id)
    return CalendarOut(start=start, end=end, days=[CalendarDayOut(**d) for d in days])


@router.get("/planning/week-print", response_model=WeekPrintOut)
def week_print(
    start: date = Query(description="Primeiro dia da semana (data local)"),
    activity_id: UUID | None = None,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> WeekPrintOut:
    """Dados de 7 dias para a versão imprimível (o frontend renderiza para impressão/PDF)."""
    return WeekPrintOut(**svc.week_print(db, user, start, activity_id=activity_id))


# --- Auto-plano ------------------------------------------------------------


@router.post("/activities/{activity_id}/auto-plan/preview", response_model=AutoPlanOut)
def auto_plan_preview(
    activity_id: UUID,
    payload: AutoPlanIn,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> AutoPlanOut:
    act = activity_service.get_activity(db, user, activity_id)
    return AutoPlanOut(
        **svc.auto_plan(
            db, user, act, start=payload.start, end=payload.end, task_ids=payload.task_ids
        )
    )


@router.post("/activities/{activity_id}/auto-plan", response_model=AutoPlanOut)
def auto_plan_apply(
    activity_id: UUID,
    payload: AutoPlanIn,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> AutoPlanOut:
    act = activity_service.get_activity(db, user, activity_id)
    res = svc.auto_plan(
        db,
        user,
        act,
        start=payload.start,
        end=payload.end,
        task_ids=payload.task_ids,
        apply=True,
    )
    audit(
        db,
        actor_id=user.id,
        action="planning.auto_plan",
        target_type="activity",
        target_id=str(act.id),
        metadata={"moved": len(res["moves"]), "unplaced": len(res["unplaced"])},
    )
    db.commit()
    return AutoPlanOut(**res)

File: apps/web/src/api/planning.ts
/** Planejamento: calendário semanal, tarefas, séries recorrentes, distribuição automática e impressão. */
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { api, unwrap } from "./client";
import { useInvalidateAll } from "./queries";
import { useUser } from "./session";
import type { components } from "./schema";

type S = components["schemas"];
export type CalendarOut = S["CalendarOut"];
export type CalendarDay = S["CalendarDayOut"];
export type Task = S["TaskOut"];
export type TaskCreate = S["TaskCreate"];
export type TaskUpdate = S["TaskUpdate"];
export type RescheduleIn = S["RescheduleIn"];
export type OccurrenceIn = S["OccurrenceIn"];
export type Series = S["SeriesOut"];
export type SeriesCreate = S["SeriesCreate"];
export type SeriesUpdate = S["SeriesUpdate"];
export type AutoPlanOut = S["AutoPlanOut"];
export type AutoPlanIn = S["AutoPlanIn"];
export type WeekPrintOut = S["WeekPrintOut"];
export type Preferences = S["PreferencesOut"];

export const planningKeys = {
  calendar: (start: string, end: string, activityId?: string | null) =>
    ["calendar", start, end, activityId ?? null] as const,
  tasks: (start: string, end: string, activityId?: string | null) =>
    ["tasks", start, end, activityId ?? null] as const,
  series: (activityId?: string | null) => ["tasks", "series", activityId ?? null] as const,
  weekPrint: (start: string) => ["planning", "week-print", start] as const,
  preferences: ["me", "preferences"] as const,
};

export function useCalendar(start: string, end: string, activityId?: string | null) {
  return useQuery({
    queryKey: planningKeys.calendar(start, end, activityId),
    queryFn: async () =>
      unwrap(
        await api.GET("/api/v1/calendar", {
          params: { query: { start, end, activity_id: activityId ?? null } },
        }),
      ) as CalendarOut,
    staleTime: 15_000,
  });
}

export function useTasks(start: string, end: string, activityId?: string | null, enabled = true) {
  return useQuery({
    queryKey: planningKeys.tasks(start, end, activityId),
    enabled,
    queryFn: async () =>
      unwrap(
        await api.GET("/api/v1/tasks", {
          params: { query: { start, end, activity_id: activityId ?? null } },
        }),
      ) as Task[],
    staleTime: 15_000,
  });
}

export function useSeries(activityId?: string | null, enabled = true) {
  return useQuery({
    queryKey: planningKeys.series(activityId),
    enabled,
    queryFn: async () =>
      unwrap(
        await api.GET("/api/v1/tasks/series", { params: { query: { activity_id: activityId ?? null } } }),
      ) as Series[],
  });
}

export function usePreferences() {
  const user = useUser();
  return useQuery({
    queryKey: planningKeys.preferences,
    enabled: !!user,
    queryFn: async () => unwrap(await api.GET("/api/v1/me/preferences")) as Preferences,
    staleTime: 5 * 60_000,
  });
}

function useInvalidatePlanning() {
  const qc = useQueryClient();
  const all = useInvalidateAll();
  return () => {
    all();
    qc.invalidateQueries({ queryKey: ["tasks"] });
    qc.invalidateQueries({ queryKey: ["planning"] });
  };
}

export function useCreateTask() {
  const invalidate = useInvalidatePlanning();
  return useMutation({
    mutationFn: async (body: TaskCreate) => unwrap(await api.POST("/api/v1/tasks", { body })) as Task,
    onSuccess: invalidate,
  });
}

export function useUpdateTask() {
  const invalidate = useInvalidatePlanning();
  return useMutation({
    mutationFn: async ({ id, body }: { id: string; body: TaskUpdate }) =>
      unwrap(await api.PATCH("/api/v1/tasks/{task_id}", { params: { path: { task_id: id } }, body })) as Task,
    onSuccess: invalidate,
  });
}

export function useDeleteTask() {
  const invalidate = useInvalidatePlanning();
  return useMutation({
    mutationFn: async (id: string) =>
      unwrap(await api.DELETE("/api/v1/tasks/{task_id}", { params: { path: { task_id: id } } })),
    onSuccess: invalidate,
  });
}

/** Reagendar mantém a mesma tarefa (id, histórico e vínculo com a série). */
export function useRescheduleTask() {
  const invalidate = useInvalidatePlanning();
  return useMutation({
    mutationFn: async ({ id, body }: { id: string; body: RescheduleIn }) =>
      unwrap(
        await api.POST("/api/v1/tasks/{task_id}/reschedule", { params: { path: { task_id: id } }, body }),
      ) as Task,
    onSuccess: invalidate,
  });
}

export function useSkipTask() {
  const invalidate = useInvalidatePlanning();
  return useMutation({
    mutationFn: async (id: string) =>
      unwrap(await api.POST("/api/v1/tasks/{task_id}/skip", { params: { path: { task_id: id } } })) as Task,
    onSuccess: invalidate,
  });
}

export function useCompleteTask() {
  const invalidate = useInvalidatePlanning();
  return useMutation({
    mutationFn: async ({ id, done }: { id: string; done: boolean }) =>
      done
        ? (unwrap(
            await api.POST("/api/v1/tasks/{task_id}/complete", { params: { path: { task_id: id } } }),
          ) as Task)
        : (unwrap(
            await api.POST("/api/v1/tasks/{task_id}/uncomplete", { params: { path: { task_id: id } } }),
          ) as Task),
    onSuccess: invalidate,
  });
}

/** Ocorrência virtual de série vira tarefa real ao ser editada/concluída. */
export function useMaterializeOccurrence() {
  const invalidate = useInvalidatePlanning();
  return useMutation({
    mutationFn: async ({ seriesId, date, body }: { seriesId: string; date: string; body: OccurrenceIn }) =>
      unwrap(
        await api.POST("/api/v1/tasks/series/{series_id}/occurrences/{local_date}", {
          params: { path: { series_id: seriesId, local_date: date } },
          body,
        }),
      ) as Task,
    onSuccess: invalidate,
  });
}

export function useCreateSeries() {
  const invalidate = useInvalidatePlanning();
  return useMutation({
    mutationFn: async (body: SeriesCreate) =>
      unwrap(await api.POST("/api/v1/tasks/series", { body })) as Series,
    onSuccess: invalidate,
  });
}

export function useUpdateSeries() {
  const invalidate = useInvalidatePlanning();
  return useMutation({
    mutationFn: async ({ id, body }: { id: string; body: SeriesUpdate }) =>
      unwrap(
        await api.PATCH("/api/v1/tasks/series/{series_id}", { params: { path: { series_id: id } }, body }),
      ) as Series,
    onSuccess: invalidate,
  });
}

export function useDeleteSeries() {
  const invalidate = useInvalidatePlanning();
  return useMutation({
    mutationFn: async (id: string) =>
      unwrap(await api.DELETE("/api/v1/tasks/series/{series_id}", { params: { path: { series_id: id } } })),
    onSuccess: invalidate,
  });
}

export function useAutoPlanPreview(activityId: string) {
  return useMutation({
    mutationFn: async (body: AutoPlanIn) =>
      unwrap(
        await api.POST("/api/v1/activities/{activity_id}/auto-plan/preview", {
          params: { path: { activity_id: activityId } },
          body,
        }),
      ) as AutoPlanOut,
  });
}

export function useAutoPlanApply(activityId: string) {
  const invalidate = useInvalidatePlanning();
  return useMutation({
    mutationFn: async (body: AutoPlanIn) =>
      unwrap(
        await api.POST("/api/v1/activities/{activity_id}/auto-plan", {
          params: { path: { activity_id: activityId } },
          body,
        }),
      ) as AutoPlanOut,
    onSuccess: invalidate,
  });
}

export function useWeekPrint(start: string, activityId?: string | null) {
  return useQuery({
    queryKey: [...planningKeys.weekPrint(start), activityId ?? null],
    queryFn: async () =>
      unwrap(
        await api.GET("/api/v1/planning/week-print", {
          params: { query: { start, activity_id: activityId ?? null } },
        }),
      ) as WeekPrintOut,
  });
}
